### DISCUSSION TITLE
Make View Transition API work with react-router

### DISCUSSION THREAD

**Discussion by lukaskoeller** [remix-run/react-router#10276] [Proposals]
## What is the new or updated feature that you are suggesting?

With Chrome 111+ comes the new [View Transition API](https://developer.chrome.com/docs/web-platform/view-transitions/) enabled in chromium browsers by default. The new feature would be to implement this API in react-router so that any added animation in CSS via `::view-transition-*` would be picked up and executed.

## Why do we need this feature?

Why view transitions are needed is well explained in this [section of the chrome article](https://developer.chrome.com/docs/web-platform/view-transitions/#why-do-we-need-this-feature).

Implementing this with react-router makes it very easy to focus on adding animations for page transitions using a native browser api. Acting / intercepting on navigation events and updating the DOM would then be part of react and react-router.
So as long as the used browser supports the view transition api, react-router would `startViewTransition` on navigation and trigger any transition if implemented in CSS.

Advantages:
* Keeps the implementation effort low, since solely the `startViewTransition` would need to be handled by the library, not the animation itself.
* It is a progressive enhancement that works if the browser supports and if not just omits this part of executes some sort a fallback.

## What is possible now?
As of now, you can already make since possible partly and with some drawbacks.

```tsx
export const CustomLink: FC<CustomLinkProps> = (props) => {
  const {
    to,
    children,
  } = props;
  const navigateTo = useNavigate();

  const navigate = (to: string) => {
    if (!document?.startViewTransition) {
      navigateTo(to);
    } else {
      document.startViewTransition(() => {
        navigateTo(to);  
      });
    }
  }

  return (
    <button onClick={() => navigate(to)}>{children}</button>
  )
}
```

This rather simple, unaccessible `<CustomLink>` component `startViewTransition` when there is an `onClick` event and then uses `navigate` to the path provided via `to`.
This works for the case of clicking the `<button>` but is not semantically correct (should be an anchor) and is not working e.g. for browser back buttons.

## Status Quo on Browser Support and Implementation

Chrome, Edge and Opera already support the View Transition API (see [can I use](https://caniuse.com/?search=View%20Transition%20API)). These two make around [70% of the worldwide browser market share](https://gs.statcounter.com/browser-market-share). [Firefox is supportive](https://github.com/mozilla/standards-positions/issues/677) about it. [Safari seems supportive](https://github.com/WebKit/standards-positions/issues/48) about it too.

* Sveltekit makes it easy to use [View Transitions with version 1.24](https://svelte.dev/blog/view-transitions) through onNavigate.
* [Astro supports View Transitions](https://docs.astro.build/en/guides/view-transitions/).
* Nuxt ships with [experimental support vor View Transitions](https://nuxt.com/docs/getting-started/transitions#view-transitions-api-experimental).
* Next has an [open discussion on adding support](https://github.com/vercel/next.js/discussions/46300)

## References
* [[v6] First-class animation primitives](https://github.com/remix-run/react-router/discussions/8008) It's core topic is animations too, but without explicitly focusing on the View Transition API.

**Comment by lili21** [comment]
we need this one.

**Comment by mrksbnch** [comment]
Similar to how [Sveltekit supports view transitions](https://svelte.dev/blog/view-transitions), it would be enough to have a lifecycle function which fires just at the right time. This would then work for all kinds of navigations (without custom links) and can also be ported over to Remix.

**Comment by yar2001** [comment]
Note that it is unreliable to put `navigate` in `document.startViewTransition` because the callback function in `startViewTransition` should be waiting until the old page dom is unmounted. Or else, the transition may start even though the old page hasn't been unmounted (especially when the new page is in suspense).

```js
document.startViewTransition(async () => {
    await oldDomUnmounted;
});
```

Therefore, a reliable solution for View Transition should be implemented internally in the React Router because the `startViewTransition` API needs to know when the new dom is finished loading (or at least the old dom is finished unmounting), and the current React Router API seems unable to achieve it conveniently.

**Comment by brophdawg11** [comment]
Good news!  Initial support for View Transitions has been added in https://github.com/remix-run/react-router/pull/10916 and will be released soon in `6.17.0` (and is currently available in `6.17.0-pre.0`).  You can find the docs on the [dev branch](https://reactrouter.com/en/dev/components/link#unstable_viewtransition) and see an example usage in https://github.com/brophdawg11/react-router-records.

**Reply by hdadr** [reply]
Whats the expected time to be shipped?

**Reply by elliotsayes** [reply]
Appears to be released already https://github.com/remix-run/react-router/blob/main/CHANGELOG.md#view-transitions-
